RedHatInsights/rhsm-subscriptions
# ENT-3618: Properly map swatch product IDs for OpenShift based on the 'product' label from Prometheus metrics.

The PR addresses 3 issues that came out of the reported bug in ENT-3618. Details can be found [here](https://issues.redhat.com/browse/ENT-3618?focusedCommentId=15901338&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15901338)

As they were all related, I've included the fixes in the same PR but 3 commits.

In summary, this PR fixes:
1) No more ProductId not found warning message for OpenShift metering products.
2) Snapshots produced from OpenShift metric events should now properly update the swatch product IDs for osd and ocp metrics.

**HOW TO TEST**
Get your token [here](https://docs.engineering.redhat.com/display/ENT/Prometheus+Metrics) and export to $PROM_TOKEN.

Run the app as follows:
```bash
$ DEV_MODE=true PROM_AUTH_TOKEN="$(echo $PROM_TOKEN)" PROM_URL="https://telemeter-lts.datahub.redhat.com/api/v1" ./gradlew clean bootRun
```

Opt-In with these 2 accounts so that you can run the tallies for each.
```bash
$ curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d  '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=optInJmxBean,type=OptInJmxBean","operation":"createOrUpdateOptInConfig(java.lang.String,java.lang.String,boolean,boolean,boolean)","arguments":["6341237","123456",true,true,true]}'

$ curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=optInJmxBean,type=OptInJmxBean","operation":"createOrUpdateOptInConfig(java.lang.String,java.lang.String,boolean,boolean,boolean)","arguments":["540155","654321",true,true,true]}'
```

Pull the metrics from Prometheus for each account.
```bash
$ curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.metering.jmx:name=openshiftJmxBean,type=OpenShiftJmxBean","operation":"performOpenshiftMeteringForAccount(java.lang.String)","arguments":["6341237"]}'

$ curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.metering.jmx:name=openshiftJmxBean,type=OpenShiftJmxBean","operation":"performOpenshiftMeteringForAccount(java.lang.String)","arguments":["540155"]}'
```

Tally the snapshots for each account.
```bash
$ curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean","operation":"tallyAccountByHourly(java.lang.String,java.lang.String,java.lang.String)","arguments":["6341237","2021-03-10T00:00:00Z","2021-03-10T23:59:59.999Z"]}'

$ curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean","operation":"tallyAccountByHourly(java.lang.String,java.lang.String,java.lang.String)","arguments":["540155","2021-03-10T00:00:00Z","2021-03-10T23:59:59.999Z"]}'
```

Verify the snapshots the product ID being used for each account.
Account 6341237 should only have product osd product: OpenShift-dedicated-metrics
Account 540155 should only have  ocp product: OpenShift-metrics
```sql
select id, account_number, snapshot_date, product_id, sla, usage, granularity from tally_snapshots s where account_number in ('6341237', '540155') order by account_number;
```
